gateway-server/src/main/java/org/apache/knox/gateway/services/security/impl/DefaultKeystoreService.java

```diff
@@ -290,6 +294,7 @@ public void addCredentialForCluster(String clusterName, String alias, String val
 
         final Path keyStoreFilePath = keyStoreDirPath.resolve(clusterName + CREDENTIALS_SUFFIX);
         writeKeyStoreToFile(ks, keyStoreFilePath, masterService.getMasterSecret());
+        addToCache(clusterName, alias, value);
```

Comment:
We want this to happen no matter what right? Maybe in the finally block or leave it at the beginning?

Reply:
I see this the same way as in the `remove` method: I would not save it in the cache unless it was saved in the keystore successfully.